### ISSUE TITLE
Delayed exception reporting in Traits

### ISSUE THREAD

**Issue by mdickinson** [enthought/traits#400]
In the session below, note that the `ValueError` doesn't appear until after pressing `<Return>` a second time after the statement that should have caused the error. This is almost certainly due to not following Python's exception protocol somewhere in ctraits.c (we're either returning NULL and not setting an exception, or vice versa).

Example session:
```
bash-4.4$ python 
Enthought Deployment Manager -- https://www.enthought.com
Python 2.7.13 |Enthought, Inc. (x86_64)| (default, Mar  2 2017, 08:20:50) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from traits.api import *
>>> class A(HasTraits):
...     foo = Either(None, Array)
... 
>>> a = A()
>>> import numpy as np
>>> a.foo = None
>>> a.foo = np.arange(5)
>>> 
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
>>> 
```

**Comment by mdickinson** [comment]
Versions: Traits 4.6.0 (from EDM), macOS 10.13.4, Python 2.7.13.

**Comment by rkern** [comment]
What's the problem? Initially, `a.foo is None` so `a.foo = None` compares `None == None`, which is fine. When you assign `a.foo = np.arange(5)`, then Traits compares `None == np.arange(5)` which causes `np.ndarray.__eq__()` to raise the `ValueError`.

**Comment by rkern** [comment]
Oh, I see.

**Comment by rkern** [comment]
I don't see this on `master`. It's possible something fixed this since the `4.6.0` release.

**Comment by rkern** [comment]
Like #377?

**Comment by rkern** [comment]
I.e. you.

**Comment by mdickinson** [comment]
@rkern Thanks, Robert. Yes, I had a suspicion that this might already be fixed in master, but didn't have the minutes available to check when I filed the report.

**Comment by mdickinson** [comment]
Closed as duplicate of #376.

**Comment by rkern** [comment]
Not an admonition. Merely amusement. :-)
